[Feat] 챌린지 공석 알림 신청, 취소 API 구현

yangjiae12 opened the pull request:
## #️⃣ 연관된 이슈

> 관련된 이슈 번호를 적어주세요.\
> Close #47 

## ✨ 작업 내용 (Summary)

> 이번 PR에서 작업한 내용을 간략히 설명해주세요. (이미지 첨부 가능)

인원이 마감된 챌린지에 공석이 생길 경우 알림을 받기 위한 공석 알림 신청, 취소 API를 구현했습니다
- `ChallengeWait` 엔티티 생성
- `ChallengeWaitRepository` 추가 및 중복 신청 방지 메서드 구현
- 챌린지 대기 신청(`register`) 및 취소(`cancel`) 비즈니스 로직 구현
- 예외 처리 로직 추가
- 중복되는 조회로직을 헬퍼 메서드로 분리

---

## ✅ 변경 사항 체크리스트

> 다음 항목들을 확인하고 체크해주세요.

- [x] 코드에 영향이 있는 모든 부분에 대한 테스트를 작성하고 실행했나요?
- [x] 문서를 작성하거나 수정했나요? (필요한 경우)
- [ ] 중요한 변경 사항이 팀에 공유되었나요?

---

## 🧪 테스트 결과

> 코드 변경에 대해 테스트를 수행한 결과를 요약해주세요.

* **테스트 환경:** f로컬
* **테스트 방법:** Swagger
* **결과 요약:** 
```
1. 정상 신청: 인원이 꽉 찬 챌린지에 대기 신청 시 성공 응답 및 DB 저장 확인
2. 정상 취소: 대기 내역 삭제 확인
3. 예외 케이스 검증
   - 이미 참여 중인 유저가 신청 시 예외 발생 확인
   - 중복 신청 시 예외 발생 확인
   - 자리가 남은 챌린지에 신청 시 예외 발생 확인
   - 추가적인 예외 케이스는 API 명세에서 확인
```
---

## 📸 스크린샷

> 관련된 스크린샷 또는 GIF가 있다면 여기에 첨부해주세요.

| 설명 | 사진 |
|---|---------|
|챌린지 공석 알림 신청|<img width="1169" height="285" alt="image" src="https://github.com/user-attachments/assets/51243756-3c96-4d61-b681-eb585df98ab2" />|
|챌린지 공석 알림 취소|<img width="1174" height="291" alt="image" src="https://github.com/user-attachments/assets/7620713c-4f37-4a32-aecc-13446eca81ec" />|
|DB 저장|<img width="1119" height="214" alt="image" src="https://github.com/user-attachments/assets/afc337c2-63c3-45ee-9223-848730b351ca" />|

---

## 💬 리뷰 요구사항

> 리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요.
- `ChallengeServiceImpl`에서 반복되는 조회 로직을 `findUser`, `findChallenge `메서드로 분리
- `void` 타입이라 `SuccessCode` 따로 분리하여 반환

---

## 📎 참고 자료

> 관련 문서, 레퍼런스 링크 등이 있다면 여기에 첨부해주세요.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **새로운 기능**
  * 모집 마감된 챌린지에 대해 대기 신청 및 대기 취소 기능 추가
  * 대기 등록/취소에 대한 성공 응답 메시지 추가

* **버그 수정**
  * 중복 참가 및 중복 대기 신청 방지 검증 강화
  * 대기 신청 조회·취소 관련 오류 처리 개선

* **변경사항**
  * 대기 관련 상태/오류 메시지 추가로 안내 강화 (등록/취소/미마감 등)

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

yangjiae12 committed:
feat: 챌린지 공석 알림 신청 및 취소를 위한 도메인 엔티티 및 리포지토리 구현 #47

yangjiae12 committed:
feat: 챌린지 공석 알림 신청 및 취소 비즈니스 로직 구현 #47

yangjiae12 committed:
feat: 챌린지 대기 신청 API 구현 및 Swagger 적용 #47

yc3697 commented on src/main/java/com/hrr/backend/domain/challenge/entity/ChallengeWait.java:
```diff
@@ -1,11 +1,36 @@
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Table(name = "challenge_wait", uniqueConstraints = {
+        @UniqueConstraint(name = "uk_challenge_wait_user_challenge", columnNames = {"user_id", "challenge_id"})
+})
 public class ChallengeWait extends BaseEntity {
+
     @Id
     @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
-}
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "user_id", nullable = false)
+    private User user;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "challenge_id", nullable = false)
+    private Challenge challenge;
+
+    @Builder
+    public ChallengeWait(User user, Challenge challenge) {
+        this.user = user;
+        this.challenge = challenge;
+    }
+}
```
최신 develop에 flyway를 도입해놔서 엔티티의 변경사항에 대해서 sql 파일 작성이 필요합니다! pull 받으셔서 작성 부탁드려요 ㅎㅎ

yangjiae12 replied:
확인했습니다~

yc3697 commented on src/main/java/com/hrr/backend/domain/challenge/service/ChallengeServiceImpl.java:
```diff
@@ -356,4 +405,16 @@ private void validateBusinessRules(ChallengeRequestDto.CreateChallengeDto req) {
+	private User findUser(Long userId) {
+		return userRepository.findById(userId)
+				.orElseThrow(() -> new GlobalException(ErrorCode.USER_NOT_FOUND));
+	}
```
@AuthenticationPrincipal CustomUserDetails을 사용했기 때문에 이미 내부적으로 User 엔티티 자체를 가지고 있어서 바로 getUser로 사용하면 DB 조회 과정을 생략할 수 있습니다. 참고해주세요

yangjiae12 replied:
확인했습니다!

yc3697 reviewed (changes requested):
리뷰 남긴 거랑 registerChallengeWait, cancelChallengeWait에 트랜잭션 어노테이션 붙이는 거 수정해주시면 좋을 것 같습니다. 수고하셨어요 ㅎㅎ

yangjiae12 committed:
Merge branch 'develop' into feat/47-challenge-wait

yangjiae12 committed:
fix: 머지 충돌 해결 시 발생한 오류 해결 #47

yangjiae12 committed:
fix: 챌린지 대기 신청 및 취소 메서드에 @Transactional 적용 #47

yangjiae12 committed:
refactor: 챌린지 관련 로직에서 불필요한 User 조회 쿼리 제거 #47

- 컨트롤러의 @AuthenticationPrincipal에서 User 객체를 바로 넘기도록 수정
- 서비스 계층의 중복된 findById 조회를 제거

yangjiae12 committed:
feat: ChallengeWait 테이블 컬럼 및 제약조건 추가 (V2.1) #47

yc3697 reviewed (approved):
확인했습니다! 마이그레이션 스크립트까지 깔끔하네요 수고하셨어요 ㅎㅎ
